### PR TITLE
pose estimation

### PR DESCRIPTION
**Pull request by tusharsangam**

### CODE REVIEW

File: bd_spot_wrapper/spot_wrapper/spot.py [hunk 1]
```diff
@@ -25,7 +25,7 @@
 import numpy as np
 import quaternion
 import rospy
-import sophus as sp
+import sophuspy as sp
```
Comment (jimmytyyang) [inline bd_spot_wrapper/spot_wrapper/spot.py:28]: this needs to be consolidate across machines

File: bd_spot_wrapper/spot_wrapper/spot.py [hunk 2]
```diff
@@ -311,6 +311,60 @@ def rotate_gripper_with_delta(self, wrist_yaw=0.0, wrist_roll=0.0):
         self.set_arm_joint_positions(new_arm_joint_states)
         time.sleep(0.5)
 
+    def move_gripper_to_points(
```
Comment (jimmytyyang) [inline bd_spot_wrapper/spot_wrapper/spot.py:314]: this is the duplication to `move_gripper_to_point`

File: bd_spot_wrapper/spot_wrapper/spot.py [hunk 3]
```diff
@@ -321,7 +375,7 @@ def move_gripper_to_point(
         :return: cmd_id
         """
         if len(rotation) == 3:  # roll pitch yaw Euler angles
-            roll, pitch, yaw = rotation
+            roll, pitch, yaw = rotation  # xyz
```
Comment (jimmytyyang) [inline bd_spot_wrapper/spot_wrapper/spot.py:378]: remove #x,y,z

File: spot_rl_experiments/experiments/skill_test/test_semantic_gaze.py [hunk 4]
```diff
@@ -2,19 +2,90 @@
...
+        input("Continue to reset arm ?")
+        spotskillmanager.get_env().reset_arm()
         contnue = map_user_input_to_boolean("Do you want to do it again ? Y/N ")
 
     # Navigate to dock and shutdown
-    spotskillmanager.dock()
+    spotskillmanager.get_env().reset_arm()
+    spotskillmanager.sit()
+
+# Backup Code
```
Comment (jimmytyyang) [inline spot_rl_experiments/experiments/skill_test/test_semantic_gaze.py:26]: we can discuss if you want to remove the following code or not when we meet

File: spot_rl_experiments/experiments/skill_test/test_semantic_gaze.py [hunk 5]
```diff
@@ -2,19 +2,90 @@
...
 
-    spotskillmanager = SpotSkillManager(use_mobile_pick=True)
+    spotskillmanager = SpotSkillManager(use_mobile_pick=False, use_semantic_place=True)
     contnue = True
+    object_name = "bottle"  # "penguin plush toy"
     while contnue:
-        spotskillmanager.semanticpick("penguin", "topdown")
-        spotskillmanager.spot.open_gripper()
+        spotskillmanager.spot.stand()
+        spotskillmanager.pick(
```
Comment (jimmytyyang) [inline spot_rl_experiments/experiments/skill_test/test_semantic_gaze.py:15]: the previous code uses the name `semanticpick` to reflect the fact that this is a newly-trained skill. Here, we are repurposing the old pick skill to have pose correction and pose estimation. Do you think we move the function to semantic pick or pick?

File: spot_rl_experiments/experiments/skill_test/test_semantic_gaze.py [hunk 6]
```diff
@@ -2,19 +2,90 @@
 # This source code is licensed under the MIT license found in the
```
Comment (jimmytyyang) [inline spot_rl_experiments/experiments/skill_test/test_semantic_gaze.py:2]: if this is different from the original semantic gaze, then I will suggest that you create a new test script

File: spot_rl_experiments/spot_rl/envs/base_env.py [hunk 7]
```diff
@@ -310,30 +313,33 @@ def step(  # noqa
...
+                    # arm_positions = np.deg2rad(self.config.PLACE_ARM_JOINT_ANGLES)
                 else:
                     self.say("BD grasp API failed.")
                     self.spot.open_gripper()
                     self.locked_on_object_count = 0
-                    arm_positions = np.deg2rad(self.config.GAZE_ARM_JOINT_ANGLES)
+                    # arm_positions = np.deg2rad(self.config.GAZE_ARM_JOINT_ANGLES)
                     time.sleep(2)
 
+                # TODO: This changes the behavior of post grasping
```
Comment (jimmytyyang) [inline spot_rl_experiments/spot_rl/envs/base_env.py:333]: I added todo here to discuss the pose grasping behavior as this substantially changes the behavior

File: spot_rl_experiments/spot_rl/envs/base_env.py [hunk 8]
```diff
@@ -458,10 +464,52 @@ def step(  # noqa
...
+                intrinsics,
+                image_src,
+                image_scale,
+                port_seg=seg_port,
+                port_pose=pose_port,
+                bbox=obj_bbox,
+                mask=mask,
+            )
+            print(f"Time taken for pose estimation {t2-t1} secs")
+            # graspmode = "topdown"
```
Comment (jimmytyyang) [inline spot_rl_experiments/spot_rl/envs/base_env.py:499]: remove

File: spot_rl_experiments/spot_rl/envs/base_env.py [hunk 9]
```diff
@@ -737,7 +785,7 @@ def get_det_bbox(self, det):
         return x1, y1, x2, y2
 
     @staticmethod
-    def locked_on_object(x1, y1, x2, y2, height, width, radius=0.15):
+    def locked_on_object(x1, y1, x2, y2, height, width, radius=0.55):
```
Comment (jimmytyyang) [inline spot_rl_experiments/spot_rl/envs/base_env.py:788]: make it to 0.15

File: spot_rl_experiments/spot_rl/envs/skill_manager.py [hunk 10]
```diff
@@ -414,7 +442,7 @@ def place(self, place_target: str = None, ee_orientation_at_grasping: np.ndarray
                 ) = detect_place_point_by_pcd_method(
                     self.spot,
                     self.pick_config.GAZE_ARM_JOINT_ANGLES,
-                    percentile=70,
+                    percentile=40,
```
Comment (jimmytyyang) [inline spot_rl_experiments/spot_rl/envs/skill_manager.py:445]: why 40?

File: spot_rl_experiments/spot_rl/skills/atomic_skills.py [hunk 11]
```diff
@@ -193,12 +193,12 @@ def execute(self, goal_dict: Dict[str, Any]) -> Tuple[bool, str]:  # noqa
         """
         status = False
         message = ""
-        try:
-            status, message = self.execute_rl_loop(goal_dict=goal_dict)
-            print(f"Feedback from skill: {message}")
-        except Exception as e:
-            message = f"Error encountered in skill execution - {e}"
-            conditional_print(message=message, verbose=self.verbose)
+        # try:
```
Comment (jimmytyyang) [inline spot_rl_experiments/spot_rl/skills/atomic_skills.py:196]: polish this

File: spot_rl_experiments/spot_rl/utils/pose_estimation.py [hunk 12]
```diff
@@ -0,0 +1,568 @@
+import time
```
Comment (jimmytyyang) [inline spot_rl_experiments/spot_rl/utils/pose_estimation.py:1]: remove the part that has large section of commaqnding

File: spot_rl_experiments/spot_rl/utils/rospy_light_detection.py [hunk 13]
```diff
@@ -0,0 +1,39 @@
+import time
```
Comment (jimmytyyang) [inline spot_rl_experiments/spot_rl/utils/rospy_light_detection.py:1]: what is this for?

File: spot_rl_experiments/spot_rl/utils/search_table_location.py [hunk 14]
```diff
@@ -406,7 +406,7 @@ def detect_place_point_by_pcd_method(
     )
 
     img_with_bbox = None
-    if visualize:
+    if True:
```
Comment (jimmytyyang) [inline spot_rl_experiments/spot_rl/utils/search_table_location.py:409]: polish this

File: spot_rl_experiments/spot_rl/utils/segmentation_service.py [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+
+    while True:
+        """A service for running segmentation service, send request using zmq socket"""
+        img, bbox = socket.recv_pyobj()
+        print("Recieved img for Segmentation")
+        masks = segment(img, np.array([bbox]), img.shape[:2], device)
+        mask = masks[0, 0].cpu().numpy()  # hxw, bool
+        socket.send_pyobj(mask)
+
+# If you want to use detection service, then use the following code to listen to socket
```
Comment (jimmytyyang) [inline spot_rl_experiments/spot_rl/utils/segmentation_service.py:127]: nice!

File: spot_rl_experiments/spot_rl/utils/pose_estimation.py [hunk 16]
```diff
@@ -0,0 +1,568 @@
...
+        correction_status, put_back_object_status = False, False
+
+        current_orientation_at_grasp_in_quat = (
+            spot.get_ee_quaternion_in_body_frame().view((np.double, 4))
+        )
+        current_point, _ = spot.get_ee_pos_in_body_frame()
+        correction_angles = self.get_correction_angle(
+            current_orientation_at_grasp_in_quat, spinal_axis
+        )
+        input(
```
Comment (jimmytyyang) [inline spot_rl_experiments/spot_rl/utils/pose_estimation.py:535]: remove

File: spot_rl_experiments/spot_rl/utils/pose_estimation.py [hunk 17]
```diff
@@ -0,0 +1,568 @@
...
+                current_grasp_orientation_name,
+                gamma,
+            )
+            if np.abs(gamma) > 30 and make_face_the_object_right
+            else current_orientation_after_grasp_in_quat
+        )
+        put_back_object_status = spot.move_gripper_to_point(
+            current_point, q_final, 10, 20
+        )
+        input("Open gripper ?")
```
Comment (jimmytyyang) [inline spot_rl_experiments/spot_rl/utils/pose_estimation.py:566]: remove

**Review by jimmytyyang — COMMENTED** [review]
Hi Tushar! I left some commands and no worries! We will discuss and then merge the PR together!!

**Comment by jimmytyyang** [comment]
- please also update the readme in spot-sim2real to give pointer about how to use pose estimation
- Have an assert that now we only support a specific mesh

**Comment by jimmytyyang** [comment]
the issue when intel camera fails

Traceback (most recent call last):
  File "/home/jmmy/research/spot-sim2real/spot_rl_experiments/experiments/skill_test/test_semantic_gaze.py", line 15, in <module>
    spotskillmanager.pick(
  File "/home/jmmy/research/spot-sim2real/spot_rl_experiments/spot_rl/envs/skill_manager.py", line 361, in pick
    status, message = self.gaze_controller.execute(goal_dict=goal_dict)
  File "/home/jmmy/research/spot-sim2real/spot_rl_experiments/spot_rl/skills/atomic_skills.py", line 197, in execute
    status, message = self.execute_rl_loop(goal_dict=goal_dict)
  File "/home/jmmy/research/spot-sim2real/spot_rl_experiments/spot_rl/skills/atomic_skills.py", line 156, in execute_rl_loop
    observations, _, done, _ = self.env.step(action_dict=action_dict)  # type: ignore
  File "/home/jmmy/research/spot-sim2real/spot_rl_experiments/spot_rl/envs/gaze_env.py", line 75, in step
    observations, reward, done, info = super().step(
  File "/home/jmmy/research/spot-sim2real/spot_rl_experiments/spot_rl/envs/base_env.py", line 316, in step
    success = self.attempt_grasp(
  File "/home/jmmy/research/spot-sim2real/spot_rl_experiments/spot_rl/envs/base_env.py", line 484, in attempt_grasp
    obj_bbox = detect_with_rospy_subscriber(object_name, image_scale)
  File "/home/jmmy/research/spot-sim2real/spot_rl_experiments/spot_rl/utils/rospy_light_detection.py", line 38, in detect_with_rospy_subscriber
    prediction = [int(float(num) / image_scale) for num in bbox_str]
UnboundLocalError: local variable 'bbox_str' referenced before assignment

**Comment by jimmytyyang** [comment]
change the default to penguin?

**Comment by jimmytyyang** [comment]
the main concerning point is the arm is not reset after calling the pick, we will discuss this later :)

File: bd_spot_wrapper/spot_wrapper/spot.py (hunk 1 above)
Comment (tusharsangam) [reply bd_spot_wrapper/spot_wrapper/spot.py:28]: I will wrap it in try-catch to make it work for both the cases

File: bd_spot_wrapper/spot_wrapper/spot.py (hunk 2 above)
Comment (tusharsangam) [reply bd_spot_wrapper/spot_wrapper/spot.py:314]: this is fixed, but we need this function

File: spot_rl_experiments/spot_rl/envs/base_env.py (hunk 7 above)
Comment (jimmytyyang) [inline spot_rl_experiments/spot_rl/envs/base_env.py:333]: remove this todo

**Review by jimmytyyang — APPROVED** [review]
LGTM!

**Pull request merged**